[LLVM][ConstantFolding] Extend constantFoldVectorReduce to include scalable vectors.

paulwalker-arm opened the pull request.

paulwalker-arm committed:
[LLVM][ConstantFolding] Extend constantFoldVectorReduce to include scalable vectors.

david-arm commented on llvm/lib/Analysis/ConstantFolding.cpp:
```diff
@@ -2163,18 +2163,39 @@ Constant *ConstantFoldBinaryFP(double (*NativeFP)(double, double),
+      return SplatVal;
+    case Intrinsic::vector_reduce_add:
+    case Intrinsic::vector_reduce_mul:
+      if (SplatVal->isZeroValue())
```
In theory you can also do a short cut for a mul reduce with a splat value of 1, if you think it's worth adding?

paulwalker-arm committed:
Remove unnecessary isa<Posion> check. Add support for vec.reduce.mul.nxv##(1).

Patch also uses isNullValue() rather than isZeroValue(), which is
less relevant for integer types.

dtcxzyw reviewed (approved):
LGTM. Thanks.
